### ISSUE TITLE
Ability to Pin model version

### ISSUE THREAD

**Issue by andreaTP** [kserve/modelmesh-serving#377]
Hi and thanks for the project!

I'm trying to pin the `versionId` of a model using the `InferenceService` CR.
The use case is to have stable and reproducible deployments even if the model in the S3 bucket gets updated.

I cannot find a relevant field in the CR, and trying to append `?versionId=<.....>` to the available fields always lead to a non-functional setup.
Is it possible to achieve this result?

**Comment by ckadner** [comment]
A model (and it's version) is uniquely identified by the storage `key` and `path`. So if you replace the model in the same bucket, same prefix, same filename, then ModelMesh will treat it as the same version of the same model. So a new model version would need a new model file path. Once the ISVC is updated there should be a new revision. 

@njhill -- do we have any good documentation on how to update an existing model with a new version and how to denote the version in the metadata? The KServe inference protocol v2 has  `model_name` and `model_version` and APIs to list/create/update/delete versions but I am not sure how it all ties together.

**Comment by tjohnson31415** [comment]
I think there may be a little confusion.
@andreaTP what you are referring to is versioning of objects with a `versionId` which is a particular feature that Amazon S3 (and probably some other S3 implementations) supports ([documentation here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html)), right?

Also ModelMesh serving mentions "S3" support, it is using that in the generic sense of services that implement the S3 API (eg. MinIO, IBM Cloud Object Storage). Use of `versionId` is new to me, but maybe support for it is already standard as part of anything that implements the "S3 API" (a quick search shows that [MinIO](https://min.io/product/object-versioning-bucket-versioning) and [IBM COS](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-versioning) do support versioning).

> Is this a deliberate design decision?

Assumed immutability of what a ISVC points to is a deliberate design decision. From the ModelMesh perspective, it does not have any explicit knowledge on the content of the S3 path, all it has is the path as the identifier. It also manages model loads/unloads dynamically and routes to all instances of the model equally. The assumption that the model is immutable makes the dynamic management of the models much simpler. I think that's the main reason for the design: it is simpler than needing a way to monitor and react to changes to objects in S3.

Not supporting `versionId` is not part of that design decision. I think we can add support for `versionId`. I would think that this would just require some changes to the S3 storage provider implementation: https://github.com/kserve/modelmesh-runtime-adapter/tree/main/pullman/storageproviders/s3

**Comment by andreaTP** [comment]
Hi @tjohnson31415 and thanks for reverting back!

I'm referring to `versionId` as the S3 feature, you are correct.

Tl;DR:
To enable basic reproducible deployments we need a way to let the user specify additional information about the content of the S3 object (`etag` , `versionId`) or, at least, verify the content against a user-provided hash (`sha`, `md5`).

The longer explanation:

> Assumed immutability of what a ISVC points to is a deliberate design decision.

This is interesting, but I want to remark on the fact that, with the current implementation, this is a **very** opinionated design decision, and should be, at least, correctly communicated in docs and fields' descriptions in the CRDs.
S3 is a mutable storage, doing assumptions over the content of an S3 object without options to specify the [`etag`](https://teppen.io/2018/06/23/aws_s3_etags/#what-is-an-s3-etag) or the `versionId` is going to result in very confusing runtime behavior.

> The assumption that the model is immutable makes the dynamic management of the models much simpler.

This is completely fair and understandable, as long as we can consider adding the necessary fields to enforce and guarantee the immutability of the models.

> I think that's the main reason for the design: it is simpler than needing a way to monitor and react to changes to objects in S3.

Again this is perfectly fine, but the current mechanism has to be improved to effectively guarantee reproducible models' deployments.

Currently, we cannot guarantee that 2 subsequent model deployments are going to serve the same model without locking to "read-only" the S3 bucket.

**Comment by njhill** [comment]
I agree with @tjohnson31415 that this seems like it would be a good thing to add, and should hopefully be a trivial change to the s3 storage provider, to recognize/use an additional optional `versionId` attribute within the storage spec.
